design/ozone.md

```diff
@@ -0,0 +1,258 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: rook-ozone
+---
+apiVersion: ozone.rook.io/v1alpha1
+kind: OzoneObjectStore
+metadata:
+  name: my-store
+  namespace: rook-ozone
+spec:
+  components: #component specific settings
+     csi:
+        enable: true
+        placement: #see rook.io/v1alpha1 PlacementSpec
+     recon:
+        enable: true
+        placement: #see rook.io/v1alpha1 PlacementSpec
+     s3g: 
+        enable: true
+        placement: #see rook.io/v1alpha1 PlacementSpec
+     scm:
+        #required component, always enabled
+        placement: #see rook.io/v1alpha1 PlacementSpec
+     om:
+        #required component, always enabled
+        placement: #see rook.io/v1alpha1 PlacementSpec
+
+  features: # these are not components but confiuration groups
+     prometheus: true
+     tracing: true
+  ozoneVersion: #see ceph versioning doc for mor details
+    image: apache/ozone:0.4.1
+    allowUnsupported: false
+  upgradePolicy: #see ceph versioning doc for mor details (Won't be implmented in Phase 1)
+    ozoneVersion:
+      image: apache/ozone:0.5.0
+      allowUnsupported: false
+  storage: #see rook.io/v1alpha1/StorageScopeSpec
+  placement: #see rook.io/v1alpha1 PlacementSpec
+  annotations: #custom annotations for all the created k8s resources
+     key: value
+  customization: #deep water: any additional transformations to modify k8s resources. Only for advanced users.
+    - type: Add
+      path: ["metadata", "labels"]
+      value:
+         generated-with: rook
+    
+```
+
+The type of the `storage` attribute is `StorageScopeSpec` which is a [common rook type](https://github.com/rook/rook/blob/818ad9c7ca4d352a9bed202e619d47b05f1a4dab/pkg/apis/rook.io/v1alpha2/types.go), and can be used to specifiy the details of the storage (eg. volumeClaimTemplates, nodeCounts, etc.)
+
+## Operator capabilities
+
+The main benefit to use an operator is to provide additional functionalities based on the current state of the runtime environment. The power of the operator based on the runtime information which is not available for any other installer tool.
+
+Some option which can be implemented as part of the Operator:
+ 
+ * Use emptyDir instead of Persistent Volumes in case of the dev cluster (_opt-in_ only, it should be requested by a CRD attribute)
+ * Connection kubernetes decommissioning and Ozone decommissioning
+ * Configure local persistence based on the available disks
+ * Support storage based auto-scaling (if datanodes are `StatefulSet`)
+ * Expose S3 endpoint based on the available ingress, LB
+ * Volume support: Ozone has a hierarchy of volumes/buckets/keys. Bucket and keys are the same what we have in any S3 compatible store, volumes can contain generic replication/security rules for all the assigned buckets. We can support Volume CRD to make it easier the management of the volumes
+ * Environment specific configuration (host names) can be adjusted based on the current topology (host names, namespace)
```

Comment:
adjusted in what ways?

Reply:
Sorry, the "adjusted" can be the wrong word. There are two kind of configuration included in `ozone-site.xml`. The runtime parameters (eg. default replication count) and the environment specific configuration (host names, ports).

The later one should be generated based on the information from the kubernetes (hostnames) even in the first version of the operator. Later it can be improved to learn the topology information (rack awareness) based on kubernetes annotations on the nodes.